### ISSUE TITLE
main-client.js vs vendor.js

### ISSUE THREAD

**Issue by aherrick** [aspnet/JavaScriptServices#1057]
Trying to get an understanding what the difference is and how these work with webpack.

Let's say I have a library I want to bring in. In this case `sweetalert2` https://github.com/limonte/sweetalert2

I then import that into my `package.json` as `    "sweetalert2": "^6.6.5"`

Now inside a component I can simply reference like so: `import swal from 'sweetalert2';` Everything works fine and `sweetalert2` is imported into my `main-client.js`

What is the need to add that to the `vendor.js` via webpack? Is that the 'main-client.js' is built up on the fly vs the `vendor.js` which is built at compile time? I must be missing something here.

**Comment by KraigWalker** [comment]
Caching is the name of the game. The idea behind having a separate `vendor.js` file is that all of your app's 3rd party dependencies can be kept apart from your own client code, with the view that your project's dependencies will change less frequently than the actual code you write for your app. 

Because the `vendor` file only needs to be recompiled when you update or add a dependency to your project, it's not updated along with the rest of the `main-client` code as to do so would be wasted effort, and increase your overall build time and feedback loop during development. 

The benefit of this can also be felt by the end-user of your app, as a cached version of `vendor.js` is more likely to stay relevant and not need to be re-downloaded when the contents of `main-client` are updated, potentially saving a lot of bandwidth.

You can use the [very nifty webpack bundle analyzer](https://www.npmjs.com/package/webpack-bundle-analyzer) plugin to check out what's being put into your different bundles - just include it as a plugin to your main-client or vendor build config and you'll be able to see the different modules being included.

It's best to move your 3rd party code (i.e npm dependencies) into your vendor file, or use Webpacks' family of chunking and module plugins to create smaller "CommonChunks" of dependencies required frequently by many components/routes.

It's even becoming a common practice to create a `runtime.js` bundle, which includes the runtime webpack itself adds to your code to handle 

I keep referring to Tim Sebastian's pretty exhaustive article on long term caching in webpack for reference. 

[https://medium.com/webpack/predictable-long-term-caching-with-webpack-d3eee1d3fa31](https://medium.com/webpack/predictable-long-term-caching-with-webpack-d3eee1d3fa31)

**Comment by aherrick** [comment]
@KraigWalker thanks for the thoughts! 

So rule of thumb being any NPM JS package I'm bringing in ensure it gets bundled as part of the `vendor.js `? (in this case I'm working with the Angular JSS webpack template setup)

Any code in my angular components will then become part of `main-client.js`?

What exactly is it then that prevents that NPM JS package from becoming part of the `main-client.js`?

Appreciate the help!

**Comment by KraigWalker** [comment]
There's nothing actually preventing npm dependencies from appearing in main-client. In fact, any imports not listed in the array in the webpack.config.vendor.js file will be placed in main-client. The vendor bundle config explicitly lists what should go into it. Then, when it is built, a json manifest file is created, which the main-client bundle uses to map imports to modules in the previously outputted vendor.js file.

The generated manifest is key to webpack being able to exclude previously compiled modules in vendor.js from appearing in main-client.js. This is all thanks to the webpack DLLPlugin.

**Comment by KraigWalker** [comment]
For what it's worth, on Monday I did a pass where we looked for files size savings in our main-client bundle, which was a monsterous 1.89mb. we found two flavours of lodash inside, the entire moment library and some redux middleware, as well as somethings we imported but never even used. After optimizong the way we imported things, and moved everything to vendor.js, our main client was just under 160kb, even less gzipped. That's a big win for us.
